etc/kernel-launchers/python/scripts/launch_ipykernel.py

```diff
@@ -360,7 +365,20 @@ def get_server_request(sock):
     return request_info
 
 
-def server_listener(sock, parent_pid):
+def cancel_spark_jobs(sig, frame):
+    try:
+        __spark_context.cancelAllJobs()
+    except Exception as e:
+        if e.__class__.__name__ == "Py4JError":
+            try:
+                __spark_context.cancelAllJobs()
+            except Exception as ex:
+                print(f"Error occurred while interrupting the kernel: {ex}")
```

Comment:
the message in L376 and L378 are same.. add something to distinguish so that incase of any issue, its easy to debug.

Reply:
I had this same comment but retracted it.  Since this makes two of us, I agree that we should slightly modify the messages so we don't need to rely on the line number info, which can be off in certain cases.

How about we change the cancellation retry attempt message to something like:
```python
except Exception as ex:
                print(f"Error occurred while re-attempting Spark job cancellation when interrupting the kernel: {ex}")
```

and the outer (_first_) attempt to something like:
```python
        else:
            print(f"Error occurred while attempting Spark job cancellation when interrupting the kernel: {e}")
```

where the _re-_ prefix is the distinguishing characteristic and we know its related to Spark job cancellation in either case.